refinery/file_store/tasks.py

```diff
@@ -20,11 +21,129 @@
+        import_file.update_state(
+            state='PROGRESS',
+            meta={
+                'percent_done': '{:.0f}'.format(percentage),
+                'current': destination_size,
+                'total': source_size
+            }
+        )
+
+
+def _copy_file(source_path, destination_path, delete_source=False):
+    """Copy file from one absolute file system path to another"""
+    _mkdir(os.path.dirname(destination_path))
+
+    try:
+        with open(source_path, 'rb') as source, \
+                open(destination_path, 'wb') as destination:
+            _copy_file_obj(source, destination)
+    except IOError as exc:
+        _delete_file(destination_path)
+        raise RuntimeError("Failed to copy '{}' to '{}': {}".format(
+            source_path, destination_path, exc
+        ))
+
+    if delete_source:
+        _delete_file(source_path)
+
+
+def _symlink_file(source_path, link_path):
+    _mkdir(os.path.dirname(link_path))
+    try:
+        os.symlink(source_path, link_path)
+    except OSError as exc:
+        raise RuntimeError("Error creating symlink '%s': %s", link_path, exc)
+
+
+def _import_path_to_path(source_path, symlink=True):
+    """Import file from an absolute file system path into file store"""
+    storage = SymlinkedFileSystemStorage()
+    source_file_name = os.path.basename(source_path)
+    file_store_name = storage.get_name(source_file_name)
+    file_store_path = storage.get_path(source_file_name)
+
+    if source_path.startswith((settings.REFINERY_DATA_IMPORT_DIR,
+                               get_temp_dir())):
+        # move uploaded or temporary file
+        logger.debug("Moving file '%s' into file store", source_path)
+        _copy_file(source_path, file_store_path, delete_source=True)
+        logger.info("Moved file '%s' into file store", source_path)
+    else:
+        if symlink:
+            _symlink_file(source_path, file_store_path)
+            logger.info("Created symlink '%s' to '%s'",
+                        file_store_path, source_path)
+        else:
+            logger.debug("Copying file '%s' into file store", source_path)
+            _copy_file(source_path, file_store_path)
+            logger.info("Copied file '%s' into file store", source_path)
+
+    return file_store_name
+
+
+def _import_path_to_s3(source_path):
+    s3 = boto3.client('s3')
+    storage = S3MediaStorage()
+    file_store_name = storage.get_name(os.path.basename(source_path))
+
+    try:
+        s3.upload_file(source_path, settings.MEDIA_BUCKET, file_store_name,
+                       ExtraArgs={'ACL': 'public-read'},
+                       Callback=ProgressPercentage(source_path,
+                                                   import_file.request.id))
+    except botocore.exceptions.ClientError as exc:
+        s3_url = 's3://' + settings.MEDIA_BUCKET + '/' + file_store_name
+        raise RuntimeError("Error transferring file '%s' to %s: %s",
```

Comment:
What if a large `source_path` is referenced here and fails right at the end? Do we leave the incomplete data in S3 or does it get removed somehow? Isn't there something tricky about multipart uploads where one can still be charged for those smaller chunks if a erring upload happens?

Reply:
This will be handled by setting an appropriate bucket policy.

Reply:
Is this bucket policy on our timeline anywhere? This will be handled in the Terraform portion of our deployment correct?

Reply:
Yes and yes.